w11h03/test/pgdp/networking/SocketTest.java

```diff
@@ -0,0 +1,455 @@
+        var in = client.getInputStream();
+        while (true) {
+            if (in.available() < 2) {
+                fail("Did not receive PartnerSwitch message (0x00, 0x04).");
+            }
+            int first = in.read();
+            if (first == 0x00) {
+                int second;
+                while (true) {
+                    second = in.read();
+                    if (second != 0x00) {
+                        break;
+                    }
+                }
+                if (second == 0x04) {
+                    break;
+                }
+            }
+        }
+    }
+
+    private static byte[] copyOf(byte[] bytes, int from, int to, String purpose) {
+        if (to < from) throw new IllegalArgumentException("`from` may not be greater than `to`.");
+        byte[] out = new byte[to - from];
+        try {
+            System.arraycopy(bytes, from, out, 0, to - from);
+        } catch (IndexOutOfBoundsException e) {
+            fail(purpose + ": Output ended unexpectedly. Tried to read from index "
+                    + from + " up to index " + to + " (exclusive).\nWas: " + Arrays.toString(bytes));
+        }
+        return out;
+    }
+
+    private void warnConnect() {
+        System.out.println("""
+                If this test doesn't finish, it's likely because you are using the address of the server (carol.sse.cit.tum.de) directly.
+                Instead, please use the variable DataHandler.serverAddress.
+                """);
+        System.out.println("""
+                If this method does not finish, you likely threw an Exception that was not DataHandler.ConnectionException or your client is trying to read bytes which are never being sent.
+                """);
+    }
+
+    private static int bufferToInt(byte[] bytes) {
+        if (bytes.length > 4) throw new IllegalArgumentException("`bytes.length` may not be larger than 4.");
+        int n = 0;
+        for (int i = 0; i < bytes.length; i++) {
+            int shift = (bytes.length - i - 1) * 8;
+            int b = bytes[i];
+            if (b < 0) b = 256 + b;
+            n |= b << shift;
+        }
+        return n;
+    }
+
+    private static long bufferToLong(byte[] bytes) {
+        if (bytes.length > 8) throw new IllegalArgumentException("`bytes.length` may not be larger than 8.");
+        long n = 0;
+        for (int i = 0; i < bytes.length; i++) {
+            int shift = (bytes.length - i - 1) * 8;
+            long b = bytes[i];
+            if (b < 0) b = 256 + b;
+            n |= b << shift;
+        }
+        return n;
+    }
+
+    /**
+     * @return All bytes sent to a socket
+     */
+    private static byte[] getAllBytes(Thread clientThread, Socket client) {
+        try {
+            clientThread.join();
+        } catch (InterruptedException ignored) {}
+        try {
+            byte[] out = new byte[client.getInputStream().available()];
+            client.getInputStream().read(out);
+            return out;
```

Comment:
This is quite error prone

Reply:
I know...
Trust me this is what I spent most of the time writing these tests on.
To my knowledge, there is no better solution, because closing the Socket and using `getAllBytes()` doesn't work.
`getAllBytes()` blocks the thread until the socket is closed. But if you close the socket (even on client using the given `close()` method, `handleInput` calls System.exit, which terminates the test thread.
If you can think of a better way to do this, please do tell.